lnwallet/fee_estimator.go

```diff
@@ -524,3 +542,205 @@ func (e ExternalFeeSource) ParseResponse(r io.Reader) (map[uint32]uint32, error)
+
+	return &WebAPIFeeEstimator{
+		apiSource:        api,
+		chainNotifier:    notifier,
+		feeByBlockTarget: make(map[uint32]uint32),
+		defaultFeePerKw:  defaultFee,
+		quit:             make(chan struct{}),
+	}
+}
+
+// EstimateFeePerKW takes in a target for the number of blocks until an initial
+// confirmation and returns the estimated fee expressed in sat/kw.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) EstimateFeePerKW(numBlocks uint32) (SatPerKWeight, error) {
+	if numBlocks > MaxBlockTarget {
+		numBlocks = MaxBlockTarget
+	} else if numBlocks < MinBlockTarget {
+		return 0, fmt.Errorf("conf target of %v is too low, minimum "+
+			"accepted is %v", numBlocks, MinBlockTarget)
+	}
+
+	w.feesMtx.Lock()
+	feePerKb, err := w.getCachedFee(numBlocks)
+	w.feesMtx.Unlock()
+	if err != nil {
+		return 0, err
+	}
+
+	// If the result is too low, then we'll clamp it to our current fee
+	// floor.
+	satPerKw := SatPerKVByte(feePerKb).FeePerKWeight()
+	if satPerKw < FeePerKwFloor {
+		satPerKw = FeePerKwFloor
+	}
+
+	walletLog.Debugf("Web API returning %v sat/kw for conf target of %v",
+		int64(satPerKw), numBlocks)
+
+	return satPerKw, nil
+}
+
+// Start signals the FeeEstimator to start any processes or goroutines it needs
+// to perform its duty.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) Start() error {
+	if !atomic.CompareAndSwapUint32(&w.started, 0, 1) {
+		return nil
+	}
+
+	if err := w.chainNotifier.Start(); err != nil {
+		return err
+	}
+
+	blockEpochs, err := w.chainNotifier.RegisterBlockEpochNtfn(nil)
+	if err != nil {
+		return err
+	}
+	w.blockEpochs = blockEpochs
+
+	w.wg.Add(1)
+	go w.blockEpochHandler()
+
+	return nil
+}
+
+// Stop stops any spawned goroutines and cleans up the resources used by the
+// fee estimator.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) Stop() error {
+	if !atomic.CompareAndSwapUint32(&w.stopped, 0, 1) {
+		return nil
+	}
+
+	close(w.quit)
+	w.wg.Wait()
+
+	return nil
+}
+
+// RelayFeePerKW returns the minimum fee rate required for transactions to be
+// relayed.
+//
+// NOTE: This method is part of the FeeEstimator interface.
+func (w *WebAPIFeeEstimator) RelayFeePerKW() SatPerKWeight {
+	return FeePerKwFloor
+}
+
+// getCachedFee takes in a target for the number of blocks until an initial
+// confirmation and returns an estimated fee (if one was returned by the API). If
+// the fee was not previously cached, we cache it here. Note that the feesMtx
+// should be held while calling this function, to avoid a race condition.
+func (w *WebAPIFeeEstimator) getCachedFee(numBlocks uint32) (uint32, error) {
+	for target := numBlocks; target >= MinBlockTarget; target-- {
+		fee, ok := w.feeByBlockTarget[target]
+		if !ok {
+			continue
+		}
+		w.feeByBlockTarget[target] = fee
```

Comment:
Style nit: missing a space here.

Reply:
This won't cache the original value, instead it'll cache the highest value that's present, which is already in the map.

Reply:
oo... fixed